File: binding/src/libqproperty.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qscopedpropertyupdategroup.html
pub const qscopedpropertyupdategroup = struct {

    /// New constructs a new QScopedPropertyUpdateGroup object.
    pub fn New() QtC.QScopedPropertyUpdateGroup {
        return qtc.QScopedPropertyUpdateGroup_new();
    }


    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QScopedPropertyUpdateGroup_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpropertybindingsourcelocation.html
pub const qpropertybindingsourcelocation = struct {

    /// New constructs a new QPropertyBindingSourceLocation object.
    pub fn New(other: ?*anyopaque) QtC.QPropertyBindingSourceLocation {
        return qtc.QPropertyBindingSourceLocation_new(@ptrCast(other));
    }


    /// New2 constructs a new QPropertyBindingSourceLocation object.
    pub fn New2(other: ?*anyopaque) QtC.QPropertyBindingSourceLocation {
        return qtc.QPropertyBindingSourceLocation_new2(@ptrCast(other));
    }


    /// New3 constructs a new QPropertyBindingSourceLocation object.
    pub fn New3() QtC.QPropertyBindingSourceLocation {
        return qtc.QPropertyBindingSourceLocation_new3();
    }


    /// New4 constructs a new QPropertyBindingSourceLocation object.
    pub fn New4(param1: ?*anyopaque) QtC.QPropertyBindingSourceLocation {
        return qtc.QPropertyBindingSourceLocation_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPropertyBindingSourceLocation_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPropertyBindingSourceLocation_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn FileName(self: ?*anyopaque, ) []const u8 {
        const fileName_ret = qtc.QPropertyBindingSourceLocation_FileName(@ptrCast(self));
return  std.mem.span(fileName_ret);
    }

    pub fn SetFileName(self: ?*anyopaque, fileName: []const u8) void {
        qtc.QPropertyBindingSourceLocation_SetFileName(@ptrCast(self), @ptrCast(fileName));
    }

    pub fn FunctionName(self: ?*anyopaque, ) []const u8 {
        const functionName_ret = qtc.QPropertyBindingSourceLocation_FunctionName(@ptrCast(self));
return  std.mem.span(functionName_ret);
    }

    pub fn SetFunctionName(self: ?*anyopaque, functionName: []const u8) void {
        qtc.QPropertyBindingSourceLocation_SetFunctionName(@ptrCast(self), @ptrCast(functionName));
    }

    pub fn Line(self: ?*anyopaque, ) u32 {
        return qtc.QPropertyBindingSourceLocation_Line(@ptrCast(self));
    }

    pub fn SetLine(self: ?*anyopaque, line: u32) void {
        qtc.QPropertyBindingSourceLocation_SetLine(@ptrCast(self), line);
    }

    pub fn Column(self: ?*anyopaque, ) u32 {
        return qtc.QPropertyBindingSourceLocation_Column(@ptrCast(self));
    }

    pub fn SetColumn(self: ?*anyopaque, column: u32) void {
        qtc.QPropertyBindingSourceLocation_SetColumn(@ptrCast(self), column);
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QPropertyBindingSourceLocation_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPropertyBindingSourceLocation_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpropertybindingerror.html
pub const qpropertybindingerror = struct {

    /// New constructs a new QPropertyBindingError object.
    pub fn New() QtC.QPropertyBindingError {
        return qtc.QPropertyBindingError_new();
    }


    /// New2 constructs a new QPropertyBindingError object.
    pub fn New2(typeVal: i32) QtC.QPropertyBindingError {
        return qtc.QPropertyBindingError_new2(@intCast(typeVal));
    }


    /// New3 constructs a new QPropertyBindingError object.
    pub fn New3(other: ?*anyopaque) QtC.QPropertyBindingError {
        return qtc.QPropertyBindingError_new3(@ptrCast(other));
    }


    /// New4 constructs a new QPropertyBindingError object.
    pub fn New4(typeVal: i32, description: []const u8) QtC.QPropertyBindingError {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
return qtc.QPropertyBindingError_new4(@intCast(typeVal), description_str);
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPropertyBindingError_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn HasError(self: ?*anyopaque, ) bool {
        return qtc.QPropertyBindingError_HasError(@ptrCast(self));
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QPropertyBindingError_Type(@ptrCast(self));
    }

    pub fn Description(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPropertyBindingError_Description(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpropertybindingerror.Description: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPropertyBindingError_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/quntypedpropertybinding.html
pub const quntypedpropertybinding = struct {

    /// New constructs a new QUntypedPropertyBinding object.
    pub fn New() QtC.QUntypedPropertyBinding {
        return qtc.QUntypedPropertyBinding_new();
    }


    /// New2 constructs a new QUntypedPropertyBinding object.
    pub fn New2(metaType: QtC.QMetaType, vtable: ?*anyopaque, function: ?*anyopaque, location: ?*anyopaque) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedPropertyBinding_new2(@ptrCast(metaType), @ptrCast(vtable), @ptrCast(function), @ptrCast(location));
    }


    /// New3 constructs a new QUntypedPropertyBinding object.
    pub fn New3(other: ?*anyopaque) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedPropertyBinding_new3(@ptrCast(other));
    }


    /// New4 constructs a new QUntypedPropertyBinding object.
    pub fn New4(priv: *QPropertyBindingPrivate) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedPropertyBinding_new4(@ptrCast(priv));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QUntypedPropertyBinding_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QUntypedPropertyBinding_IsNull(@ptrCast(self));
    }

    pub fn Error(self: ?*anyopaque, ) QtC.QPropertyBindingError {
        return qtc.QUntypedPropertyBinding_Error(@ptrCast(self));
    }

    pub fn ValueMetaType(self: ?*anyopaque, ) QtC.QMetaType {
        return qtc.QUntypedPropertyBinding_ValueMetaType(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUntypedPropertyBinding_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpropertyobserverbase.html
pub const qpropertyobserverbase = struct {

    /// New constructs a new QPropertyObserverBase object.
    pub fn New(other: ?*anyopaque) QtC.QPropertyObserverBase {
        return qtc.QPropertyObserverBase_new(@ptrCast(other));
    }


    /// New2 constructs a new QPropertyObserverBase object.
    pub fn New2(other: ?*anyopaque) QtC.QPropertyObserverBase {
        return qtc.QPropertyObserverBase_new2(@ptrCast(other));
    }


    /// New3 constructs a new QPropertyObserverBase object.
    pub fn New3() QtC.QPropertyObserverBase {
        return qtc.QPropertyObserverBase_new3();
    }


    /// New4 constructs a new QPropertyObserverBase object.
    pub fn New4(param1: ?*anyopaque) QtC.QPropertyObserverBase {
        return qtc.QPropertyObserverBase_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPropertyObserverBase_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPropertyObserverBase_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPropertyObserverBase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpropertyobserver.html
pub const qpropertyobserver = struct {

    /// New constructs a new QPropertyObserver object.
    pub fn New() QtC.QPropertyObserver {
        return qtc.QPropertyObserver_new();
    }


    pub fn SetSource(self: ?*anyopaque, property: ?*anyopaque) void {
        qtc.QPropertyObserver_SetSource(@ptrCast(self), @ptrCast(property));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPropertyObserver_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpropertynotifier.html
pub const qpropertynotifier = struct {

    /// New constructs a new QPropertyNotifier object.
    pub fn New() QtC.QPropertyNotifier {
        return qtc.QPropertyNotifier_new();
    }


    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPropertyNotifier_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/quntypedbindable.html
pub const quntypedbindable = struct {

    /// New constructs a new QUntypedBindable object.
    pub fn New(other: ?*anyopaque) QtC.QUntypedBindable {
        return qtc.QUntypedBindable_new(@ptrCast(other));
    }


    /// New2 constructs a new QUntypedBindable object.
    pub fn New2(other: ?*anyopaque) QtC.QUntypedBindable {
        return qtc.QUntypedBindable_new2(@ptrCast(other));
    }


    /// New3 constructs a new QUntypedBindable object.
    pub fn New3() QtC.QUntypedBindable {
        return qtc.QUntypedBindable_new3();
    }


    /// New4 constructs a new QUntypedBindable object.
    pub fn New4(param1: ?*anyopaque) QtC.QUntypedBindable {
        return qtc.QUntypedBindable_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QUntypedBindable_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QUntypedBindable_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QUntypedBindable_IsValid(@ptrCast(self));
    }

    pub fn IsBindable(self: ?*anyopaque, ) bool {
        return qtc.QUntypedBindable_IsBindable(@ptrCast(self));
    }

    pub fn IsReadOnly(self: ?*anyopaque, ) bool {
        return qtc.QUntypedBindable_IsReadOnly(@ptrCast(self));
    }

    pub fn MakeBinding(self: ?*anyopaque, ) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedBindable_MakeBinding(@ptrCast(self));
    }

    pub fn TakeBinding(self: ?*anyopaque, ) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedBindable_TakeBinding(@ptrCast(self));
    }

    pub fn Observe(self: ?*anyopaque, observer: ?*anyopaque) void {
        qtc.QUntypedBindable_Observe(@ptrCast(self), @ptrCast(observer));
    }

    pub fn Binding(self: ?*anyopaque, ) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedBindable_Binding(@ptrCast(self));
    }

    pub fn SetBinding(self: ?*anyopaque, binding: ?*anyopaque) bool {
        return qtc.QUntypedBindable_SetBinding(@ptrCast(self), @ptrCast(binding));
    }

    pub fn HasBinding(self: ?*anyopaque, ) bool {
        return qtc.QUntypedBindable_HasBinding(@ptrCast(self));
    }

    pub fn MetaType(self: ?*anyopaque, ) QtC.QMetaType {
        return qtc.QUntypedBindable_MetaType(@ptrCast(self));
    }

    pub fn MakeBinding1(self: ?*anyopaque, location: ?*anyopaque) QtC.QUntypedPropertyBinding {
        return qtc.QUntypedBindable_MakeBinding1(@ptrCast(self), @ptrCast(location));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUntypedBindable_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/property.html#types
pub const enums = struct {
    pub const Reason = enum {
        pub const InvalidInterface: i32 = 0;
        pub const NonBindableInterface: i32 = 1;
        pub const ReadOnlyInterface: i32 = 2;
    };

    pub const Type = enum {
        pub const NoError: i32 = 0;
        pub const BindingLoop: i32 = 1;
        pub const EvaluationError: i32 = 2;
        pub const UnknownError: i32 = 3;
    };

    pub const ObserverTag = enum {
        pub const ObserverNotifiesBinding: i32 = 0;
        pub const ObserverNotifiesChangeHandler: i32 = 1;
        pub const ObserverIsPlaceholder: i32 = 2;
    };

};

File: binding/src/libqproperty.cpp
#include <QMetaType>
#include <QPropertyBindingError>
#include <QPropertyBindingSourceLocation>
#include <QPropertyNotifier>
#include <QPropertyObserver>
#include <QPropertyObserverBase>
#include <QScopedPropertyUpdateGroup>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUntypedBindable>
#include <QUntypedPropertyBinding>
#include <QUntypedPropertyData>
#include <qproperty.h>
#include "libqproperty.h"
#include "libqproperty.hxx"

QScopedPropertyUpdateGroup* QScopedPropertyUpdateGroup_new() {
	 return new QScopedPropertyUpdateGroup();
}

void QScopedPropertyUpdateGroup_Delete(QScopedPropertyUpdateGroup* self) {
    delete self;
}

QPropertyBindingSourceLocation* QPropertyBindingSourceLocation_new(const QPropertyBindingSourceLocation* other) {
	 return new QPropertyBindingSourceLocation(*other);
}

QPropertyBindingSourceLocation* QPropertyBindingSourceLocation_new2(QPropertyBindingSourceLocation* other) {
	 return new QPropertyBindingSourceLocation(*other);
}

QPropertyBindingSourceLocation* QPropertyBindingSourceLocation_new3() {
	 return new QPropertyBindingSourceLocation();
}

QPropertyBindingSourceLocation* QPropertyBindingSourceLocation_new4(const QPropertyBindingSourceLocation* param1) {
	 return new QPropertyBindingSourceLocation(*param1);
}

void QPropertyBindingSourceLocation_CopyAssign(QPropertyBindingSourceLocation* self, QPropertyBindingSourceLocation* other) {
    *self = *other;
}

void QPropertyBindingSourceLocation_MoveAssign(QPropertyBindingSourceLocation* self, QPropertyBindingSourceLocation* other) {
    *self = std::move(*other);
}

const char* QPropertyBindingSourceLocation_FileName(const QPropertyBindingSourceLocation* self) {
	return self->fileName;
}

void QPropertyBindingSourceLocation_SetFileName(QPropertyBindingSourceLocation* self, const char* fileName) {
	self->fileName;
}

const char* QPropertyBindingSourceLocation_FunctionName(const QPropertyBindingSourceLocation* self) {
	return self->functionName;
}

void QPropertyBindingSourceLocation_SetFunctionName(QPropertyBindingSourceLocation* self, const char* functionName) {
	self->functionName;
}

unsigned int QPropertyBindingSourceLocation_Line(const QPropertyBindingSourceLocation* self) {
	return self->line;
}

void QPropertyBindingSourceLocation_SetLine(QPropertyBindingSourceLocation* self, unsigned int line) {
	self->line;
}

unsigned int QPropertyBindingSourceLocation_Column(const QPropertyBindingSourceLocation* self) {
	return self->column;
}

void QPropertyBindingSourceLocation_SetColumn(QPropertyBindingSourceLocation* self, unsigned int column) {
	self->column;
}

void QPropertyBindingSourceLocation_OperatorAssign(QPropertyBindingSourceLocation* self, const QPropertyBindingSourceLocation* param1) {
	self->operator=(*param1);
}

void QPropertyBindingSourceLocation_Delete(QPropertyBindingSourceLocation* self) {
    delete self;
}

QPropertyBindingError* QPropertyBindingError_new() {
	 return new QPropertyBindingError();
}

QPropertyBindingError* QPropertyBindingError_new2(int typeVal) {
	 return new QPropertyBindingError(static_cast<QMetaType::Type>(typeVal));
}

QPropertyBindingError* QPropertyBindingError_new3(const QPropertyBindingError* other) {
	 return new QPropertyBindingError(*other);
}

QPropertyBindingError* QPropertyBindingError_new4(int typeVal, const libqt_string description) {
	 return new QPropertyBindingError(static_cast<QMetaType::Type>(typeVal), QString::fromUtf8(description.data, description.len));
}

void QPropertyBindingError_OperatorAssign(QPropertyBindingError* self, const QPropertyBindingError* other) {
	self->operator=(*other);
}

bool QPropertyBindingError_HasError(const QPropertyBindingError* self) {
	return self->hasError();
}

int QPropertyBindingError_Type(const QPropertyBindingError* self) {
	return self->type();
}

libqt_string QPropertyBindingError_Description(const QPropertyBindingError* self) {
	QString _ret = self->description();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPropertyBindingError_Delete(QPropertyBindingError* self) {
    delete self;
}

QUntypedPropertyBinding* QUntypedPropertyBinding_new() {
	 return new QUntypedPropertyBinding();
}

QUntypedPropertyBinding* QUntypedPropertyBinding_new2(const QUntypedPropertyBinding* other) {
	 return new QUntypedPropertyBinding(*other);
}

void QUntypedPropertyBinding_OperatorAssign(QUntypedPropertyBinding* self, const QUntypedPropertyBinding* other) {
	self->operator=(*other);
}

bool QUntypedPropertyBinding_IsNull(const QUntypedPropertyBinding* self) {
	return self->isNull();
}

QPropertyBindingError* QUntypedPropertyBinding_Error(const QUntypedPropertyBinding* self) {
	return new QPropertyBindingError(self->error());
}

QMetaType* QUntypedPropertyBinding_ValueMetaType(const QUntypedPropertyBinding* self) {
	return new QMetaType(self->valueMetaType());
}

void QUntypedPropertyBinding_Delete(QUntypedPropertyBinding* self) {
    delete self;
}

QPropertyObserverBase* QPropertyObserverBase_new(const QPropertyObserverBase* other) {
	 return new QPropertyObserverBase(*other);
}

QPropertyObserverBase* QPropertyObserverBase_new2(QPropertyObserverBase* other) {
	 return new QPropertyObserverBase(*other);
}

QPropertyObserverBase* QPropertyObserverBase_new3() {
	 return new QPropertyObserverBase();
}

QPropertyObserverBase* QPropertyObserverBase_new4(const QPropertyObserverBase* param1) {
	 return new QPropertyObserverBase(*param1);
}

void QPropertyObserverBase_CopyAssign(QPropertyObserverBase* self, QPropertyObserverBase* other) {
    *self = *other;
}

void QPropertyObserverBase_MoveAssign(QPropertyObserverBase* self, QPropertyObserverBase* other) {
    *self = std::move(*other);
}

void QPropertyObserverBase_Delete(QPropertyObserverBase* self) {
    delete self;
}

QPropertyObserver* QPropertyObserver_new() {
	 return new QPropertyObserver();
}

void QPropertyObserver_Delete(QPropertyObserver* self) {
    delete self;
}

QPropertyNotifier* QPropertyNotifier_new() {
	 return new QPropertyNotifier();
}

void QPropertyNotifier_Delete(QPropertyNotifier* self) {
    delete self;
}

QUntypedBindable* QUntypedBindable_new(const QUntypedBindable* other) {
	 return new QUntypedBindable(*other);
}

QUntypedBindable* QUntypedBindable_new2(QUntypedBindable* other) {
	 return new QUntypedBindable(*other);
}

QUntypedBindable* QUntypedBindable_new3() {
	 return new QUntypedBindable();
}

QUntypedBindable* QUntypedBindable_new4(const QUntypedBindable* param1) {
	 return new QUntypedBindable(*param1);
}

void QUntypedBindable_CopyAssign(QUntypedBindable* self, QUntypedBindable* other) {
    *self = *other;
}

void QUntypedBindable_MoveAssign(QUntypedBindable* self, QUntypedBindable* other) {
    *self = std::move(*other);
}

bool QUntypedBindable_IsValid(const QUntypedBindable* self) {
	return self->isValid();
}

bool QUntypedBindable_IsBindable(const QUntypedBindable* self) {
	return self->isBindable();
}

bool QUntypedBindable_IsReadOnly(const QUntypedBindable* self) {
	return self->isReadOnly();
}

QUntypedPropertyBinding* QUntypedBindable_MakeBinding(const QUntypedBindable* self) {
	return new QUntypedPropertyBinding(self->makeBinding());
}

QUntypedPropertyBinding* QUntypedBindable_TakeBinding(QUntypedBindable* self) {
	return new QUntypedPropertyBinding(self->takeBinding());
}

void QUntypedBindable_Observe(const QUntypedBindable* self, QPropertyObserver* observer) {
	self->observe(observer);
}

QUntypedPropertyBinding* QUntypedBindable_Binding(const QUntypedBindable* self) {
	return new QUntypedPropertyBinding(self->binding());
}

bool QUntypedBindable_SetBinding(QUntypedBindable* self, const QUntypedPropertyBinding* binding) {
	return self->setBinding(*binding);
}

bool QUntypedBindable_HasBinding(const QUntypedBindable* self) {
	return self->hasBinding();
}

QMetaType* QUntypedBindable_MetaType(const QUntypedBindable* self) {
	return new QMetaType(self->metaType());
}

QUntypedPropertyBinding* QUntypedBindable_MakeBinding1(const QUntypedBindable* self, const QPropertyBindingSourceLocation* location) {
	return new QUntypedPropertyBinding(self->makeBinding(*location));
}

void QUntypedBindable_Delete(QUntypedBindable* self) {
    delete self;
}

